core/shared/src/main/scala/org/http4s/Entity.scala

```diff
@@ -19,20 +19,45 @@ package org.http4s
 import cats.Monoid
 import cats.syntax.all._
 import cats.~>
+import fs2.Pure
+import fs2.Stream
 
-final case class Entity[+F[_]](body: EntityBody[F], length: Option[Long] = None) {
-  def translate[F1[x] >: F[x], G[_]](fk: F1 ~> G): Entity[G] = Entity(body.translate(fk), length)
+sealed trait Entity[+F[_]] {
+  def body: EntityBody[F]
+  def length: Option[Long]
+  def ++[F1[x] >: F[x]](that: Entity[F1]): Entity[F1]
+  def translate[F1[x] >: F[x], G[_]](fk: F1 ~> G): Entity[G]
 }
 
 object Entity {
+  def apply[F[_]](body: EntityBody[F], length: Option[Long] = None): Entity[F] =
+    Default(body, length)
+
+  // The type parameter aids type inference in Message constructors. This should be unnecessary when Messages are covariant.
+  def empty[F[_]]: Entity[F] = Empty
+
+  final case class Default[+F[_]](body: EntityBody[F], length: Option[Long]) extends Entity[F] {
+    def ++[F1[x] >: F[x]](that: Entity[F1]): Entity[F1] = that match {
+      case d: Default[F1] => Default(body ++ d.body, (length, d.length).mapN(_ + _))
+      case Empty => this
+    }
+
+    def translate[F1[x] >: F[x], G[_]](fk: F1 ~> G): Entity[G] = Default(body.translate(fk), length)
+  }
+
+  case object Empty extends Entity[Pure] {
```

Comment:
Something quite nice about this :)

Reply:
We could do a type alias I suppose.

Reply:
The problem is that `Pure` is defined by an inequality, `Pure[A] <: Nothing`. So if we define our own one it won't be the same as fs2's one, unless it's just an alias for `fs2.Pure`.